libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py

```diff
@@ -60,19 +79,320 @@ def validate_environment(  # pylint: disable=no-self-argument
+
         values["endpoint"] = get_from_dict_or_env(
             values,
             "endpoint",
             "OCI_LLM_ENDPOINT",
         )
         return values
 
+    def _headers(self, is_async=False, body=None) -> Dict:
+        """Construct and return the headers for a request.
+
+        Args:
+            is_async (bool, optional): Indicates if the request is asynchronous.
+                Defaults to `False`.
+            body (optional): The request body to be included in the headers if
+                the request is asynchronous.
+
+        Returns:
+            Dict: A dictionary containing the appropriate headers for the request.
+        """
+        if is_async:
+            signer = self.auth["signer"]
+            req = requests.Request("POST", self.endpoint, json=body)
+            req = req.prepare()
+            req = signer(req)
+            headers = {}
+            for key, value in req.headers.items():
+                headers[key] = value
+
+            if self.streaming:
+                headers.update(
+                    {"enable-streaming": "true", "Accept": "text/event-stream"}
+                )
+            return headers
+
+        return (
+            {
+                "Content-Type": DEFAULT_CONTENT_TYPE_JSON,
+                "enable-streaming": "true",
+                "Accept": "text/event-stream",
+            }
+            if self.streaming
+            else {
+                "Content-Type": DEFAULT_CONTENT_TYPE_JSON,
+            }
+        )
+
+    def completion_with_retry(
+        self, run_manager: Optional[CallbackManagerForLLMRun] = None, **kwargs: Any
+    ) -> Any:
+        """Use tenacity to retry the completion call."""
+        retry_decorator = _create_retry_decorator(self, run_manager=run_manager)
+
+        @retry_decorator
+        def _completion_with_retry(**kwargs: Any) -> Any:
+            try:
+                request_timeout = kwargs.pop("request_timeout", DEFAULT_TIME_OUT)
+                data = kwargs.pop("data")
+                stream = kwargs.pop("stream", self.streaming)
+
+                request = Requests(
+                    headers=self._headers(), auth=self.auth.get("signer")
+                )
+                response = request.post(
+                    url=self.endpoint,
+                    data=data,
+                    timeout=request_timeout,
+                    stream=stream,
+                    **kwargs,
+                )
+                self._check_response(response)
+                return response
+            except TokenExpiredError as e:
+                raise e
+            except Exception as err:
+                logger.debug(
+                    f"Requests payload: {data}. Requests arguments: "
+                    f"url={self.endpoint},timeout={request_timeout},stream={stream}."
+                    f"Additional request kwargs={kwargs}."
+                )
+                raise ValueError(
```

Comment:
Maybe instead of using ```ValueError``` we can create a custom error?

```
class InferenceEndpointError(Exception):
    """Custom exception for errors occurring at the inference endpoint."""
    def __init__(self, message: str, original_exception: Exception) -> None:
        super().__init__(message)
        self.original_exception = original_exception
```


```
raise InferenceEndpointError(
                f"Error occurred at inference endpoint: {str(err)}", err
            ) from err
```

Reply:
I used `RuntimeError` at beginning, then switch back to `ValueError` as I have observed that all other provider using this name.  If `ValueError` is not ok, then probably `RuntimeError` can be a better name.

Reply:
I'm ok with either. However, using ```ValueError``` for general exceptions might not be the best approach, especially if the error does not specifically relate to an invalid value.

Reply:
Sure. Changed to `RuntimeError`.